filebeat/input/filestream/fswatch.go

```diff
@@ -326,93 +358,134 @@ func (s *fileScanner) GetFiles() map[string]os.FileInfo {
+	it.info, err = os.Lstat(it.filename) // to determine if it's a symlink
 	if err != nil {
-		s.log.Debugf("finding path to original file has failed %s: %+v", file, err)
-		return true
+		return it, fmt.Errorf("failed to lstat %q: %w", it.filename, err)
 	}
-	// Check if original file is included to make sure we are not reading from
-	// unwanted files.
-	if s.isFileExcluded(originalFile) || !s.isFileIncluded(originalFile) {
-		s.log.Debugf("Exclude original file: %s", file)
-		return true
+
+	if it.info.IsDir() {
+		return it, fmt.Errorf("file %q is a directory", it.filename)
 	}
 
-	return false
+	it.symlink = it.info.Mode()&os.ModeSymlink > 0
+
+	if it.symlink {
+		if !s.cfg.Symlinks {
+			return it, fmt.Errorf("file %q is a symlink and they're disabled", it.filename)
+		}
+
+		// now we know it's a symlink, we stat with link resolution
+		it.info, err = os.Stat(it.filename)
+		if err != nil {
+			return it, fmt.Errorf("failed to stat the symlink %q: %w", it.filename, err)
+		}
+
+		it.originalFilename, err = filepath.EvalSymlinks(it.filename)
+		if err != nil {
+			return it, fmt.Errorf("failed to resolve the symlink %q: %w", it.filename, err)
+		}
+
+		if s.isFileExcluded(it.originalFilename) {
+			return it, fmt.Errorf("file %q->%q is excluded from ingestion", it.filename, it.originalFilename)
+		}
+
+		if !s.isFileIncluded(it.originalFilename) {
+			return it, fmt.Errorf("file %q->%q is not included in ingestion", it.filename, it.originalFilename)
+		}
+	}
+
+	return it, nil
 }
 
-func (s *fileScanner) isOriginalAndSymlinkConfigured(file string, uniqFileID map[string]os.FileInfo) bool {
-	if s.symlinks {
-		fileInfo, err := os.Stat(file)
+func (s *fileScanner) toFileDescriptor(it *ingestTarget) (fd loginp.FileDescriptor, err error) {
+	fd.Filename = it.filename
+	fd.Info = it.info
+
+	if s.cfg.Fingerprint.Enabled {
+		fileSize := it.info.Size()
+		minSize := s.cfg.Fingerprint.Offset + s.cfg.Fingerprint.Length
+		if fileSize < minSize {
+			return fd, fmt.Errorf("filesize of %q is %d, expected at least %d for fingerprinting", fd.Filename, fileSize, minSize)
+		}
+
+		h := sha256.New()
+		file, err := os.Open(it.originalFilename)
 		if err != nil {
-			s.log.Debugf("stat(%s) failed: %s", file, err)
-			return false
+			return fd, fmt.Errorf("failed to open %q for fingerprinting: %w", it.originalFilename, err)
 		}
-		fileID := file_helper.GetOSState(fileInfo).String()
-		if finfo, exists := uniqFileID[fileID]; exists {
-			s.log.Infof("Same file found as symlink and original. Skipping file: %s (as it same as %s)", file, finfo.Name())
-			return true
+		defer file.Close()
+
+		if s.cfg.Fingerprint.Offset != 0 {
+			_, err = file.Seek(s.cfg.Fingerprint.Offset, io.SeekStart)
+			if err != nil {
+				return fd, fmt.Errorf("failed to seek %q for fingerprinting: %w", fd.Filename, err)
+			}
+		}
+
+		r := io.LimitReader(file, s.cfg.Fingerprint.Length)
+		buf := make([]byte, h.BlockSize())
+		_, err = io.CopyBuffer(h, r, buf)
```

Comment:
Corner case, but should probably verify the full return value here in case the previous size metadata is out of sync with the file on disk -- should get an error if `CopyBuffer` returns an error _or_ we copied less than `Fingerprint.Length` bytes

Reply:
Good point, thanks!

Reply:
Fixed in https://github.com/elastic/beats/pull/35734/commits/11ff1890629f64ca612b2737835fae02190b5d3c